kong/core/balancer.lua

```diff
@@ -309,6 +310,7 @@ local function execute(target)
 
   target.ip = ip
   target.port = port
+  target.hostname = target.host
```

Comment:
On retries, the balancer is called again: https://github.com/Mashape/kong/blob/189628b28c672886b1d74de9bf1857c1c7a80719/kong/kong.lua#L213-L222